File: c/quicksort.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void quickSort(int v[], int n);
void swap(int v[], int i, int j);
int scmp(const void *p1, const void *p2);

int main() {
  int v[] = {
    376, 990, 99, 78, 334, 343,
    376, 990, 99, 78, 334, 343,
    376, 990, 99, 78, 334, 343
  };
  int sizeofAll = (int)sizeof(v);
  int sizeofOne = (int)sizeof(v[0]);
  int n = sizeofAll / sizeofOne;
  printf("sizeAll = %d, sizeOne = %d, n = %d\n", sizeofAll, sizeofOne, n);
  quickSort(v, n);
  for (int i = 0; i < n; i++) {
    printf("%d ", v[i]);
  }
  printf("\n");
  return 0;
}

void quickSort(int v[], int n) {
  int i, last;
  if (n <= 1) {
    return;
  }
  swap(v, 0, rand() % n);
  last = 0;
  for (i = 1; i < n; i++) {
    if (v[i] < v[0]) {
      swap(v, ++last, i);
    }
  }
  swap(v, 0, last);
  quickSort(v, last);
  quickSort(v+last+1, n-last-1);
}

void swap(int v[], int i, int j) {
  int temp = v[i];
  v[i] = v[j];
  v[j] = temp;
}

/* scmp: string compare of *p1 and *p2 */
int scmp(const void *p1, const void *p2) {
  char *v1, *v2;
  v1 = *(char **) p1;
  v2 = *(char **) p2;
  return strcmp(v1, v2);
}

File: javascript/quicksort.js

// 快速排序算法：实现升序排列
// @param <必传 inputs> Array 待排序数据，数组类型
function quickSort(inputs) {
  if (inputs.length <= 1) {
    return
  }
  sort(inputs)
}

// 快速排序算法：内部实现
// @param <必传 inputs> Array 待排序数据，数组类型
// @param <非必 start> 待排序子数组的起始位置，递归时传入
// @param <非必 end> 待排序子数组的结束位置，递归时传入
function sort(inputs, start, end) {
  var x = start === undefined ? 0 : start
  var y = end === undefined ? inputs.length - 1 : end
  var n = y - x + 1
  if (n <= 1) {
    return
  }

  swap(inputs, x, rand(x, y)) // 选择 pivot

  var last = 0
  for (var i = 1; i < n; i++) { // 进行 partition
    if (inputs[x+i] < inputs[x]) {
      swap(inputs, x+(++last), x+i)
    }
  }
  var idx = x + last
  swap(inputs, x, idx) // 恢复 pivot
  sort(inputs, x, idx-1)
  sort(inputs, idx+1, y)
}

// 交换函数：辅助功能
function swap(arr, i, j) {
  if ( i === j) {
    return
  }
  var temp = arr[i]
  arr[i] = arr[j]
  arr[j] = temp
}

// 生成指定范围随机整数
function rand(min, max) {
  var range = max - min
  var rand = Math.random()
  return min + Math.round(rand * range)
}
